waiter/src/waiter/reporter.clj

```diff
@@ -87,15 +100,53 @@
                s/Any s/Any}
               (merge {:pickled? true} config)))
 
+(defn- make-reporting-name
+  "Concatenates elements to form a dotted name for reporting, eliding any null values or empty strings."
+  [prefix & args]
+  (MetricRegistry/name prefix (into-array String args)))
+
+(let [df (DecimalFormat. "#.####################")]
+  (defn- format-value
+    "Convert numbers to string to send to graphite server"
+    [value]
+    (if (number? value)
+      (.format df value)
+      (str value))))
+
+(defn- report-to-graphite-helper
+  "Recursively traverse metrics map and send to graphite"
+  [prefix map ^GraphiteSender graphite timestamp]
+  (if (map? map)
+    (doseq [[k v] map]
+      (report-to-graphite-helper (make-reporting-name prefix (str k)) v graphite timestamp))
```

Comment:
Nitpick: if we do stay with the recursive solution, we can use `recur` here to make it clearer that we are recursing here.
`(recur (make-reporting-name prefix (str k)) v graphite timestamp)`